misk/src/main/kotlin/misk/eventrouter/RealEventRouter.kt

```diff
@@ -4,80 +4,170 @@ import com.google.common.cache.CacheBuilder
 
-  /** We queue up subscribes when we don't have a cluster. */
-  private fun initCluster() {
-    while (subscriptionQueue.isNotEmpty()) {
-      val queuedSubscribe = subscriptionQueue.pop()
-      subscribe(queuedSubscribe)
-    }
+  internal fun drainQueue() {
+    if (!hasClusterSnapshot.get()) return
+
+    while (true) {
+      val action = actionQueue.poll() ?: return
 
-    while (publishQueue.isNotEmpty()) {
-      val (topic, event) = publishQueue.pop()
-      publish(topic, event)
+      logger.debug { "[${clusterSnapshot.self}]: processing action: $action" }
+      when (action) {
+        is Action.LeaveCluster -> clusterConnector.leaveCluster(topicPeer)
+
+        is Action.ClusterChanged -> {
+          val removedHosts = clusterSnapshot.hosts.minus(action.newSnapshot.hosts)
+          for (host in removedHosts) {
+            logger.debug { "[${clusterSnapshot.self}]: invalidating $host" }
+            hostToSocket.invalidate(host)
+          }
+
+          clusterSnapshot = action.newSnapshot
+        }
+
+        is Action.OnMessage -> {
+          val socketEvent = eventJsonAdapter.fromJson(action.text)!!
+          when (socketEvent) {
+            is SocketEvent.Event -> {
+              localSubscribers.get(socketEvent.topic).forEach {
+                (it.listener as Listener<String>).onEvent(it as Subscription<String>,
+                    socketEvent.message)
+              }
+
+              remoteSubscribers.get(socketEvent.topic).forEach {
+                it.send(action.text)
+              }
+            }
+
+            is SocketEvent.Subscribe -> {
+              remoteSubscribers.put(socketEvent.topic, action.webSocket)
+              action.webSocket.send(eventJsonAdapter.toJson(SocketEvent.Ack(socketEvent.topic)))
+            }
+
+            is SocketEvent.Ack -> {
+              localSubscribers.get(socketEvent.topic).forEach {
+                it.onOpen()
+              }
+            }
+
+            is SocketEvent.Unsubscribe -> {
+              remoteSubscribers.get(socketEvent.topic).remove(action.webSocket)
+            }
+          }
+        }
+
+        is Action.Subscribe -> {
+          val topicName = action.localSubscription.topic.name
+          val topicOwner = clusterMapper.topicToHost(clusterSnapshot, topicName)
+          if (topicOwner != clusterSnapshot.self) {
+            val subscribeEvent = eventJsonAdapter.toJson(SocketEvent.Subscribe(topicName))
+            hostToSocket[topicOwner].send(subscribeEvent)
+          } else {
+            action.localSubscription.onOpen()
+          }
+
+          localSubscribers.put(topicName, action.localSubscription)
+        }
+
+        is Action.Publish -> {
+          val topicOwner = clusterMapper.topicToHost(clusterSnapshot, action.topic)
+          val socketEvent = SocketEvent.Event(action.topic, action.event.toString())
+          val eventJson = eventJsonAdapter.toJson(socketEvent)
+
+          if (topicOwner != clusterSnapshot.self) {
```

Comment:
love it

Reply:
(I’m wondering about redirecting publishes, and the consequences of infinite loops)

Reply:
we should write some sweet tests for that